File: Problems/ShortestWordDistanceII.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class WordDistance
{
public:
    WordDistance(vector<string> &wordsDict)
    {
        for (int i = 0; i < wordsDict.size(); i++)
            this->umap[wordsDict[i]].push_back(i);
    }

    int shortest(string word1, string word2)
    {
        vector<int> list1 = this->umap[word1];
        vector<int> list2 = this->umap[word2];
        int res = INT_MAX, i = 0, j = 0;
        while (i < list1.size() && j < list2.size())
        {
            res = min(res, abs(list1[i] - list2[j]));
            list1[i] < list2[j] ? i++ : j++;
        }
        return res;
    }

private:
    unordered_map<string, vector<int>> umap;
};

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance* obj = new WordDistance(wordsDict);
 * int param_1 = obj->shortest(word1,word2);
 */

File: Problems/ShortestWordDistanceII.py
from collections import defaultdict
from typing import List


class WordDistance:

    def __init__(self, wordsDict: List[str]):
        self.dic = defaultdict()
        for i, word in enumerate(wordsDict):
            if word in self.dic:
                self.dic[word].append(i)
            else:
                self.dic[word] = []
                self.dic[word].append(i)

    def shortest(self, word1: str, word2: str) -> int:
        list1 = self.dic[word1]
        list2 = self.dic[word2]
        res = sys.maxsize
        for index1 in list1:
            for index2 in list2:
                res = min(res, abs(index1 - index2))
        return res


# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(wordsDict)
# param_1 = obj.shortest(word1,word2)
